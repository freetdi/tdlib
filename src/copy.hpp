//TODO: header

#ifndef TREEDEC_COPY_HPP
#define TREEDEC_COPY_HPP

#include <boost/graph/copy.hpp>

#ifdef HAVE_GALA_GRAPH_H
#include <gala/boost_copy.h>
#endif

namespace treedec{

using boost::num_vertices;

// copy s to t, but weed out multiedges
// rationale: when boost::copying a directed graph (0,1),(1,0) to an empty
// undirected multigraph, then we get two edges, while one might be more useful
// and more efficient sometimes.
template<class S, class T /* class M */>
void copy_trace(const S& s, T& t /*, M map=identity */)
{
    // TODO template metamagic...
    if(boost::num_vertices(t)){ untested();
        assert(false);
    }else if(!boost::is_multigraph<T>()){
        boost::copy_graph(s, t);
    }else if(boost::is_directed(t) && !boost::is_bidirectional_graph<T>::value) {
        boost::copy_graph(s, t);
    }else{
        t = MOVE(T(boost::num_vertices(s)));
        auto b=boost::edges(s);
        for(; b.first!=b.second; ++b.first){
            typename boost::graph_traits<S>::edge_descriptor f = *b.first;
            auto u=boost::source(f, s);
            auto v=boost::target(*b.first, s);
            if(u<v){
                boost::add_edge(u, v, t);
            }else{
            }
        }
    }
}

} //namespace treedec

#endif

// vim:ts=8:sw=4:et:
