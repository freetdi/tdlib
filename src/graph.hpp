// Felix Salfelder, 2015 - 2016
//
// (c) 2016 Goethe-Universit√§t Frankfurt
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//

/*
 * Graph routines related to tree decompositions.
 *
 * Implemented for generic BGL graphs. Possibly (much) faster for particular
 * graphs.
 *
 * template<typename nIter_t, typename G_t>
 *    void make_clique(nIter_t nIter, G_t &G)
 *
 * template<typename B_t, typename nIter_t, typename G_t>
 *    void copy_neighbourhood(B_t &B, nIter_t nIter, G_t &G)
 *
 * template<class G>
 * void detach_neighborhood(typename boost::graph_traits<G>::vertex_descriptor& c,
 *    G& g, typename outedge_set<G>::type& bag)
 */

#ifndef TD_GRAPH_H
#define TD_GRAPH_H

#include <boost/graph/adjacency_list.hpp>

#include "degree.hpp"
#include "error.hpp"
#include "graph_traits.hpp"
#include "treedec_traits.hpp"
#include "platform.hpp"
#include "random_generators.hpp"
#include "trace.hpp"
#include "error.hpp"

#ifndef HAVE_MYBOOL
#define HAVE_MYBOOL
class mybool{ //
public:
	mybool() : value_(bool())
	{
	}
	/* explicit */ mybool(bool const& t): value_(t) {}
	// /* explicit */ operator bool&() { return value_; }
	/* explicit */ operator bool() const { return value_; }
private:
	char value_;
};
#endif

namespace treedec{ //

#define vertex_iterator_G typename boost::graph_traits<G>::vertex_iterator
#define vertex_descriptor_G typename boost::graph_traits<G>::vertex_descriptor
#define adjacency_iterator_G typename boost::graph_traits<G>::adjacency_iterator

template<typename G>
void remove_vertex(vertex_iterator_G u, G &g)
{
    remove_vertex(*u, g);
}

//Vertex v will remain as isolated node.
//Calls cb on neighbors if degree drops by one,
//before dg will drop.
// TODO: pass edge or edge iterator.
template<typename G>
void contract_edge(vertex_descriptor_G v,
                   vertex_descriptor_G target,
                   G &g,
                   bool /*erase*/=false,
                   vertex_callback<vertex_descriptor_G>* cb=NULL)
{
    adjacency_iterator_G I, E;
    for(boost::tie(I, E)=boost::adjacent_vertices(v, g); I!=E; ++I){
        assert(boost::edge(v, *I, g).second);
        if(*I != target){
            bool added=boost::add_edge(target, *I, g).second;
            if(added){
                //rebasing edge from I-v to I-target.
            }else if(cb){
                //did not add, degree will drop by one.
                (*cb)(*I);
            }
        }else{
        }
    }

    boost::clear_vertex(v, g);
    assert(!boost::edge(v, target, g).second);
}

//Vertex v will remain as isolated node, unless erase.
//If erase, v will be deleted (not collapsed).
template<typename G>
inline void contract_edge(vertex_iterator_G v,
                   vertex_descriptor_G into,
                   G &g,
                   bool erase=false,
                   vertex_callback<vertex_descriptor_G>* cb=NULL)
{ untested();
    contract_edge(*v, into, g, erase, cb);
    if(erase){untested();
        remove_vertex(v, g);
    }
}

} // noboost

namespace treedec{
// turn vertex range into clique.
// call cb on newly created edges and incident vertices.
// returns the number of newly created edges.
template<typename B, typename E, typename G_t>
size_t // hmm
//typename boost::graph_traits<G_t>::edges_size_type
make_clique(B nIt1, E nEnd, G_t &G, typename treedec::graph_callback<G_t>* cb=NULL)
{
    typedef typename boost::graph_traits<G_t>::edge_descriptor edge_descriptor;
    size_t counter=0;
    B nIt2;
    for(; nIt1!=nEnd; ++nIt1){
#if 1
        if(cb){
            (*cb)(*nIt1); // hmm
        }else{
        }
#endif
        nIt2 = nIt1;
        nIt2++;
        for(; nIt2 != nEnd; nIt2++){
            std::pair<edge_descriptor, bool> ep=boost::add_edge(*nIt1, *nIt2, G);
            if(ep.second){
               // new edge created
               ++counter;

               if(cb){
#if 0 // doesn't work. removing center...?
            (*cb)(*nIt1); // hmm
            (*cb)(*nIt2); // hmm
#endif
                   (*cb)(*nIt1, *nIt2);
               }else{
               }
            }else{
            }
            if(!boost::is_directed(G)){
            }else if( boost::edge(*nIt2, *nIt1, G).second ){
                // change later.
            }else{
                boost::add_edge(*nIt2, *nIt1, G);
            }
        }
    }
    return counter;
}

template<typename nIter_t, typename G_t>
void make_clique(nIter_t nIter, G_t &G, treedec::graph_callback<G_t>* cb=NULL)
{
    typename boost::graph_traits<G_t>::adjacency_iterator nIt1, nEnd;
    boost::tie(nIt1, nEnd) = nIter;
    make_clique(nIt1, nEnd, G, cb);
}

// insert the neighbors of v in G into B
template<typename B_t, typename V_t, typename G_t>
void insert_neighbours(B_t &B, V_t v, G_t const &G)
{
    typename boost::graph_traits<G_t>::adjacency_iterator nIt, nEnd;
    boost::tie(nIt, nEnd) = boost::adjacent_vertices(v, G);
    for(; nIt!=nEnd; ++nIt){
        B.insert(*nIt);
    }
}

template<typename B_t, typename V_t, typename G_t>
void insert_neighbours(B_t &B, V_t v, V_t w, G_t const &G)
{
    insert_neighbours(B, v, G);
    insert_neighbours(B, w, G);
}

template<typename B_t, typename V_t, typename G_t>
void insert_neighbours(B_t &B, V_t v, V_t w, V_t x, G_t const &G)
{ untested();
    insert_neighbours(B, v, w, G);
    insert_neighbours(B, x, G);
}

// insert the neighbors of v in G into B
// B starts empty.
// equivalent to B = outedge_set(v, G) where applicable
template<typename B_t, typename V_t, typename G_t>
void assign_neighbours(B_t &B, V_t v, G_t const &G)
{
    typename boost::graph_traits<G_t>::adjacency_iterator nIt, nEnd;
    boost::tie(nIt, nEnd) = boost::adjacent_vertices(v, G);
    for(; nIt!=nEnd; ++nIt){
        B.insert(B.end(), *nIt);
    }
}

// insert neighbors of both vertices, after clearing B
template<typename B_t, typename V_t, typename G_t>
void assign_neighbours(B_t &B, V_t v, V_t w, V_t x, G_t const &G)
{
    B.clear();
    insert_neighbours(B, v, w, G);
    insert_neighbours(B, x, G);
}

// find a minimum degree vertex using linear search.
template <typename G_t>
inline typename boost::graph_traits<G_t>::vertex_descriptor
   get_min_degree_vertex(const G_t &G, bool ignore_isolated_vertices=false)
{
    unsigned int min_degree = UINT_MAX;

    typename boost::graph_traits<G_t>::vertex_iterator vIt, vEnd;
    boost::tie(vIt, vEnd) = boost::vertices(G);
    typename boost::graph_traits<G_t>::vertex_descriptor min_vertex = *vIt++;
    for(; vIt != vEnd; vIt++){
        unsigned int degree = boost::degree(*vIt, G);
        if(degree <= min_degree){
            if(ignore_isolated_vertices && degree == 0){ continue; }
            min_degree = degree;
            min_vertex = *vIt;
        }
    }

    return min_vertex;
}

template <typename G_t>
inline typename boost::graph_traits<G_t>::vertex_descriptor
   get_least_common_vertex(const typename boost::graph_traits<G_t>::vertex_descriptor &min_vertex,
           const G_t &G);

//Copy vertices of G into degree_sequence, ordered by degree, starting with lowest.
template <typename G_t>
inline void make_degree_sequence(const G_t &G,
          std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> &degree_sequence)
{
    unsigned int max_degree = 0;
    typename boost::graph_traits<G_t>::vertex_iterator vIt, vEnd;
    for(boost::tie(vIt, vEnd) = boost::vertices(G); vIt != vEnd; vIt++){
        unsigned int degree = boost::degree(*vIt, G);
        max_degree = (degree>max_degree)? degree : max_degree;
    }

    std::vector<std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> > buckets(max_degree+1);
    for(boost::tie(vIt, vEnd) = boost::vertices(G); vIt != vEnd; vIt++){
        unsigned int degree = boost::degree(*vIt, G);
        if(degree > 0){
            buckets[degree].push_back(*vIt);
        }
    }
    for(unsigned int i = 1; i <= max_degree; i++){
        for(unsigned int j = 0; j < buckets[i].size(); j++){
            degree_sequence.push_back(buckets[i][j]);
        }
    }
}

// obsolete?
template<class G>
struct outedge_set;

template <typename G_t>
std::pair<typename boost::graph_traits<typename graph_traits<G_t>::directed_overlay>::vertex_descriptor,
          typename boost::graph_traits<typename graph_traits<G_t>::directed_overlay>::vertex_descriptor>
    make_digraph_with_source_and_sink(G_t const &G, std::vector<bool> const &disabled,
                 unsigned num_dis,
                 typename graph_traits<G_t>::directed_overlay& dg,
                 std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> &idxMap,
                 typename std::set<typename boost::graph_traits<G_t>::vertex_descriptor> const &X,
                 typename std::set<typename boost::graph_traits<G_t>::vertex_descriptor> const &Y);


#if 0 // not in development?
namespace detail{ //
template<class B, class T, class V>
struct tmpbaghack{ //
    static typename treedec_traits<T>::bag_type& get_bag(T& t, V& v)
    {
        return t[v];
    }
    static typename treedec_traits<T>::bag_type const& get_bag(T const& t, V const& v)
    {
        return t[v];
    }
};

template<class T_t, class V>
struct tmpbaghack<bag_t, T_t, V>{ //
    static typename treedec_traits<T_t>::bag_type& get_bag(T_t& t, V& v)
    {
        return t[v].bag;
    }
    static typename treedec_traits<T_t>::bag_type const& get_bag(T_t const& t, V const& v)
    {
        return t[v].bag;
    }
};
} //detail

template<typename T_t>
inline typename treedec_traits<T_t>::bag_type& bag(
	const typename boost::graph_traits<T_t>::vertex_descriptor& v,
        T_t& T)
{
    typedef typename T_t::vertex_property_type b; //>::bag_type b;
    return detail::tmpbaghack<b,T_t,const typename boost::graph_traits<T_t>::vertex_descriptor&>::get_bag(T, v);
}

template<typename T_t>
inline typename treedec_traits<T_t>::bag_type const& bag(
        const typename boost::graph_traits<T_t>::vertex_descriptor& v,
        T_t const& T)
{
    typedef typename T_t::vertex_property_type b; //>::bag_type b;
    return detail::tmpbaghack<b,T_t,const typename boost::graph_traits<T_t>::vertex_descriptor&>::get_bag(T, v);
}

template<class V, class G>
size_t bag_size(V const & v, G const& g)
{
    return bag(v, g).size();
}
#endif

template<class G_t>
struct deg_chooser { //
    typedef typename misc::DEGS<G_t> type;
    typedef type degs_type; // transition? don't use.
};

// put neighbors of c into bag
// isolate c in g
// return bag
// optionally: pass pointer to bag for storage.
//TODO: not here.
template<class G, class B>
inline void detach_neighborhood(
        typename boost::graph_traits<G>::vertex_descriptor& c,
        G& g, B& bag);

// count number of edges missing in 1-neighborhood of v
template <typename G_t>
inline size_t count_missing_edges(
        const typename boost::graph_traits<G_t>::vertex_descriptor v, G_t const &G)
{
    size_t missing_edges = 0;

    typename boost::graph_traits<G_t>::adjacency_iterator nIt1, nIt2, nEnd;
    for(boost::tie(nIt1, nEnd) = boost::adjacent_vertices(v, G); nIt1 != nEnd; nIt1++){
        nIt2 = nIt1;
        nIt2++;
        for(; nIt2 != nEnd; nIt2++){
            if(!boost::edge(*nIt1, *nIt2, G).second){
                ++missing_edges;
            }
        }
    }
    return missing_edges;
}

// collect neighbors of c into bag
// remove c from graph (i.e. isolate).
// turn subgraph induced by bag into clique.
//
// optionally:
// - call back on each newly created edge, and on each vertex incident to such
//   edge.
// - provide memory through bag pointer.
//
template<class G, class B>
size_t /*G::edge_index_type?*/ make_clique_and_detach(
        typename boost::graph_traits<G>::vertex_descriptor c,
        G& g, B& bag,
        treedec::graph_callback<G>* cb=NULL)
{
    if(boost::is_undirected(g)){
        detach_neighborhood(c, g, bag);
        return make_clique(bag.begin(), bag.end(), g, cb);
    }else{
        /// directed. leave center node untouched.
        // this is a vile hack
        // (and slow for boost::adj_list)
        auto b=boost::adjacent_vertices(c, g);
        return make_clique(b.first, b.second, g, cb);

        auto p=boost::adjacent_vertices(c, g);
        for(; p.first!=p.second; ++p.first)
        {
            // inefficient...
            boost::remove_edge(*p.first, c, g);
        }
    }
}

namespace detail{ //

    // iterate over edges adjacent to both v and s
    // implementation: iterate outedges(v), skip non-outedges(s).
    template<class G>
    class shared_adj_iter : public boost::graph_traits<G>::adjacency_iterator{ //
    public:
        typedef typename boost::graph_traits<G>::adjacency_iterator adjacency_iterator;
        typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
    public: // construct
        shared_adj_iter(adjacency_iterator v, adjacency_iterator ve,
                        vertex_descriptor s, G const& g)
            : adjacency_iterator(v), _ve(ve),
              _s(s), _g(g)
        {
            skip();
        }
        shared_adj_iter(vertex_descriptor v,
                        vertex_descriptor s, G const& g)
            : adjacency_iterator(boost::adjacent_vertices(v, g).first),
              _ve(boost::adjacent_vertices(v, g).second),
              _s(s), _g(g)
        {untested();
            skip();
        }
        shared_adj_iter(const shared_adj_iter& p)
            : adjacency_iterator(p), _ve(p._ve), _s(p._s), _g(p._g)
        {
        }
    public: //ops
        shared_adj_iter& operator++(){
            assert(_ve!=adjacency_iterator(*this));
            assert(adjacency_iterator(*this)!=_ve);
            adjacency_iterator::operator++();
            skip();

            return *this;
        }
    private:
        void skip()
        {
            while(true){
                if(typename boost::graph_traits<G>::adjacency_iterator(*this)==_ve){
                    return;
                }else if(!boost::edge(**this, _s, _g).second){
                    adjacency_iterator::operator++();
                }else{
                    return;
                }
            }
        }
        adjacency_iterator _ve;
        vertex_descriptor _s;
        G const& _g;
    };

}// detail

template<class G>
std::pair<detail::shared_adj_iter<G>, detail::shared_adj_iter<G> >
    inline common_out_edges(typename boost::graph_traits<G>::vertex_descriptor v,
                     typename boost::graph_traits<G>::vertex_descriptor w,
                     const G& g)
{
    typedef typename detail::shared_adj_iter<G> Iter;
    typedef typename boost::graph_traits<G>::adjacency_iterator adjacency_iterator;

    std::pair<adjacency_iterator, adjacency_iterator> p=boost::adjacent_vertices(v, g);
    return std::make_pair(Iter(p.first, p.second, w, g),
                          Iter(p.second, p.second, w, g));
}

// check if two vertices have the same neighborhood.
// return false if not.
template<class G_t>
bool check_twins(typename boost::graph_traits<G_t>::vertex_descriptor v,
                 typename boost::graph_traits<G_t>::vertex_descriptor w,
                 const G_t& G)
{
    typename graph_traits<G_t>::outedge_set_type N1, N2;
    assign_neighbours(N1, v, G);
    assign_neighbours(N2, w, G);

    return(N1==N2);
}

// create ig.
// edge(v, w, ig) == edge(ordering[v], ordering[w], g)
// TODO: inverse map?
// TODO: partial/induced graph?
template<class G>
void immutable_clone(G const &g, typename graph_traits<G>::immutable_type& ig,
   std::vector<typename boost::graph_traits<G>::vertex_descriptor>* ordering=NULL)
{
    typedef typename graph_traits<G>::immutable_type immutable_type;
//    typedef typename graph_traits<G>::vertex_iterator vertex_iterator;
    typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
    typedef typename boost::graph_traits<G>::edge_iterator edge_iterator;

    if(!ordering){ incomplete();
    }
    assert(ordering->size()==boost::num_vertices(g)); // for now.


    if(boost::num_vertices(g) != boost::num_vertices(ig)){ untested();
        // drop a new one... (for now?)
        // inefficient.
        ig = MOVE(immutable_type(boost::num_vertices(g)));
    }else if(boost::num_edges(ig)){ untested();
        for(unsigned i=0; i<boost::num_vertices(g); ++i){
            boost::clear_vertex(i, ig);
        }
    }else{ untested();
    }

    std::vector<unsigned> inverse_ordering(boost::num_vertices(g));
    for(unsigned o=0; o<boost::num_vertices(g); ++o){
        assert((*ordering)[o] < inverse_ordering.size());
        inverse_ordering[(*ordering)[o]] = o;
    }

    for(unsigned o=0; o<boost::num_vertices(g); ++o){
        assert((*ordering)[inverse_ordering[o]] == o);
    }

    edge_iterator e, eend;
    boost::tie(e,eend) = boost::edges(g);
    for(;e!=eend; ++e){
        vertex_descriptor s = boost::source(*e, g);
        vertex_descriptor t = boost::target(*e, g);
        assert(inverse_ordering[s] < boost::num_vertices(ig));
        assert(inverse_ordering[t] < boost::num_vertices(ig));
        assert((s==t) == (inverse_ordering[s]==inverse_ordering[t]));
        boost::add_edge(inverse_ordering[s], inverse_ordering[t], ig);
    }
}

// check if {*vd1, *vd2} is a subset of a bag adjacent to t.
template<class VD_t, class T_t>
class is_in_neighbour_bd{ //
public:
    is_in_neighbour_bd(T_t const& T,
        typename boost::graph_traits<T_t>::vertex_descriptor t)
       : _T(T), _t(t)
    {
    }
public:
    bool operator() (VD_t vd1, VD_t vd2)
    {
        assert(vd1!=vd2);
        if(vd1<vd2){
        }else{ untested();
        }

        typedef typename boost::graph_traits<T_t>::adjacency_iterator bag_iterator;
        bag_iterator nIt, nEnd;
        boost::tie(nIt, nEnd) = boost::adjacent_vertices(_t, _T);
        for(; nIt != nEnd; nIt++){
            BOOST_AUTO(const& ibag, bag(*nIt, _T));

            // BUG, does not work on vectors.
            if(ibag.find(vd1)==ibag.end()){
            }else if(ibag.find(vd2)==ibag.end()){
            }else{
                return true;
            }
        }
        // a = vd1;
        // b = vd2;
        return false;
    }
private: //data
    T_t const &_T;
    typename boost::graph_traits<T_t>::vertex_descriptor _t;
public: // HACK
    long unsigned a, b;
};

// clone subgraph induced by bag into ig.
// store map V(H) -> X \subset V(G) in vdMap
// add more edges for MSVS, TODO: implement properly (how?)
template<class G_t, class T_t, class IG_t, class M_t>
inline void induced_subgraph_with_extra_edges
(G_t const &G, IG_t& ig, // typename graph_traits<G>::immutable_type& ig,
       T_t const& T, typename boost::graph_traits<T_t>::vertex_descriptor bd,
    //   URGHS. no default types without c++11.
     M_t* vdMap /*=NULL*/);


// insert reverse edges.
// dont know how to express "orientation" yet
template<class G>
void make_symmetric(G&g, bool force_oriented=false)
{ untested();
    // assert(g.is_directed);
    auto EE=boost::edges(g);
    if ( /*g.oriented ||*/ force_oriented) { untested();
        for(;EE.first!=EE.second; ++EE.first){
            auto s=boost::source(*EE.first, g);
            auto t=boost::target(*EE.first, g);
            boost::add_edge(t,s,g);
        }
    }else{ untested();
        for(;EE.first!=EE.second; ++EE.first){
            auto s=boost::source(*EE.first, g);
            auto t=boost::target(*EE.first, g);
            if(!boost::edge(s,t,g).second){ untested();
                unreachable(); //?
                boost::add_edge(s,t,g);
            }else if(!boost::edge(t,s,g).second){ untested();
                boost::add_edge(t,s,g);
            }else{ untested();
            }
        }
    }
}
} // treedec


#endif // guard
// vim:ts=8:sw=4:et
