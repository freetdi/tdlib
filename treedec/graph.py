# (c) 2021 Felix Salfelder
# GPLv3+

from . import _graph
from ._graph import _balvvd
