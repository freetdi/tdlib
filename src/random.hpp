// Lukas Larisch, 2014 - 2016
//
// (c) 2014-2016 Goethe-Universit√§t Frankfurt
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//

#ifndef TD_RANDOM
#define TD_RANDOM

#include <vector>
#include <algorithm>    // std::random_shuffle

#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/copy.hpp>

#ifdef _OPENMP
#include <omp.h>
#endif

namespace treedec{

namespace random{

template <typename G_t>
int try_some_elimination_orderings(G_t &G,
       std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> &best_elim_ordering,
       unsigned int count = 10)
{
    std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> elim_ordering(boost::num_vertices(G));
    for(unsigned int i=0; i<elim_ordering.size(); ++i){
        elim_ordering[i] = i;
    }

    std::vector<std::vector<typename boost::graph_traits<G_t>::vertex_descriptor> > elimination_orderings(count);
    for(unsigned int i=0; i<elimination_orderings.size(); ++i){
        //Use the built-in random generator.
        std::random_shuffle(elim_ordering.begin(), elim_ordering.end());
        elimination_orderings[i] = elim_ordering;
    }

    int min_width = INT_MAX;

#ifdef OPENMP
    #pragma omp parallel for
#endif
    for(unsigned int i = 0; i < count; i++){
        G_t H;
        boost::copy_graph(G, H);
        int width_i = treedec::get_width_of_elimination_ordering(H, elimination_orderings[i]);
        if(width_i < min_width){
            min_width = width_i;
            best_elim_ordering = MOVE(elimination_orderings[i]);
        }
    }

    return min_width;
}

} //namespace random

} //namespace treedec

#endif //TD_RANDOM
