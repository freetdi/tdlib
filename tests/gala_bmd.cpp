
#ifdef HAVE_GALA
#include <gala/boost.h>
#include <gala/graph.h>
#include <gala/td.h>
#include <gala/trace.h>
#endif

#include <boost/graph/copy.hpp>
#include <boost/graph/random.hpp>
#include <boost/random.hpp>
#include <iostream>
#include <stdlib.h>
#include <tdlib/elimination_orderings.hpp>
#include <tdlib/graph.hpp>
#include <tdlib/minimum_degree_ordering.hpp>
#include <tdlib/trace.hpp>
// not in development yet
//#include <tdlib/printer.hpp>


#ifdef HAVE_GALA
template<class G>
struct dvv_config : public gala::graph_cfg_default<G>
{
	static constexpr bool is_directed = true;
};

typedef uint32_t unsignedType;

#if 1
typedef gala::graph<std::vector, std::vector, unsignedType, dvv_config> sg_dvv;
#else
typedef gala::graph<std::set, std::vector, uint16_t, dvv_config> sg_dvv;
#endif

#endif

typedef boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS> bald_t;
typedef boost::adjacency_list<boost::setS, boost::vecS, boost::undirectedS> balu_t;

// typedef bald_t parsegr_t;

int main(int argc, char** argv)
{
#ifndef HAVE_GALA
	(void) argc;
	(void) argv;
	return 77;
#else

#if 0
	PARSE* p;
	try{
		p = new PARSE(std::cin);
	}catch(...){
		std::cout << "uuh\n";
		exit(2);
	}
#endif

	size_t size=1<<5;
	if(argc>1) size=1<<atoi(argv[1]);
	size_t ne=size*size/8 + 1;

	boost::mt19937 rng;
	if(argc>2){
		rng.seed(atoi(argv[2]));
	}
	bool print=false;
	if(argc>3){
		print = true;
	}

	typedef sg_dvv G;
	sg_dvv g(size);
	g.reshape(0);
	boost::generate_random_graph(g, size, ne, rng);
	size_t e=boost::num_edges(g);
	size_t n=boost::num_vertices(g);
	std::cout << "generated " << e << " edges, " << n << " vertices\n";
	treedec::check(g);
	g.make_symmetric(false);
	treedec::check(g);
	e = boost::num_edges(g);

	// boost::add_edge(0,1,g);
	e=boost::num_edges(g);

	std::cout << "symmetric " << e << " edges\n";


	unsigned i=0;
	auto E=boost::edges(g);
	for(;E.first!=E.second; ++E.first){
		++i;

		std::cout << boost::source(*E.first, g) << " -- " <<
			boost::target(*E.first, g) << "\n";

		if(i==5) break;

	}

	std::deque<unsigned long > iso;
	auto V=boost::vertices(g);
	for(;V.first!=V.second; ++V.first){
			if(boost::out_degree(*V.first, g)){

			}else{
				iso.push_back(*V.first);
			}
	}

#if 0 // incomplete
	(vec_to_treedec now works anyway!)
	for(auto i: iso){
		boost::remove_vertex(i, g);
	}
#endif

	n = boost::num_vertices(g);

	V=boost::vertices(g);
	for(;V.first!=V.second; ++V.first){
		assert(*V.first<n);
		// assert(boost::out_degree(*V.first,g));
	}

	std::cout << "removed " << iso.size() <<"\n";

	i = 0;
	// boost md does not like cliques.
	trace3("clique check", n, e, size);
	if((n*(n-1u)) == boost::num_edges(g)){ untested();
		exit(0);
	}else{
		itested();
	}

	std::vector<int> inverse_perm(n, 0);
	std::vector<int> supernode_sizes(n, 1);
	auto id = boost::get(boost::vertex_index, g);
	std::vector<int> degree(n, 0);
	std::vector<int> io(n, 0);
	std::vector<int> o(n, 0);

	G h(g);
	assert(boost::num_edges(g)==boost::num_edges(g));

	// assert(n + iso.size() == size);

	/*
	 * (Graph& g,
	 *  DegreeMap degree,
	 *  InversePermutationMap inverse_perm,
	 *  PermutationMap perm,
	 *  SuperNodeMap supernode_size,
	 *  int delta,
	 *  VertexIndexMap vertex_index_map)
	 */

	unsigned ub=-1;

	unsigned w =
#ifndef HAVE_MINDEGREE_FORK
		0;
	itested();
#endif
	boost::minimum_degree_ordering
		(g,
		 boost::make_iterator_property_map(&degree[0], id, degree[0]),
		 &io[0],
		 &o[0],
		 boost::make_iterator_property_map(&supernode_sizes[0], id, supernode_sizes[0]),
		 0,
		 id
#ifdef HAVE_MINDEGREE_FORK
		 , ub
#endif
		);
	typename treedec::graph_traits<G>::treedec_type t;
	g=h; // restore
	treedec::check(g);
	assert(boost::num_edges(g)==boost::num_edges(h));

	std::vector<bool> permcheck(n, false);
	for(auto x : o){
		permcheck[x]=true;
	}
	for(auto x : permcheck){
		assert(x);
	}

	int status = 0;
#ifdef CHECK_INEFFICIENT_VARIANT
	treedec::ordering_to_treedec(g, o, t ); // can kill g!
   h=g; // restore

	status = treedec::check_treedec(g,t);
	std::cout << "bagsize " << treedec::get_bagsize(t) << " status " << status <<"\n";
	std::cout << "same as bagsize! " << w <<"\n";
	assert(w == treedec::get_bagsize(t)); /// checks if BMD works!
#endif

	std::vector<unsigned>& ou=reinterpret_cast<std::vector<unsigned>&>(o);
	treedec::draft::vec_ordering_to_tree(g, ou, t );
	assert(boost::num_edges(t)+1==boost::num_vertices(t));
	assert(boost::num_edges(t)+1==boost::num_vertices(g));

	if(print){
		auto P = treedec::grtdprinter<G>(std::cout, g, "test");
		treedec::draft::vec_ordering_to_tree(g, ou, P );
	}


	// does not work on vectors...
	// status=treedec::check_treedec(g,t);

	if (!status) std::cout << "treedec is valid!!\n";
	std::cout << "bagsize " << treedec::get_bagsize(t) << " status " << status <<"\n";
	std::cout << "boost_minDegree_ordering said " << w << "\n";

	incomplete();
//	assert(!status);

	if(w != treedec::get_bagsize(t)){
	}
	assert(w == treedec::get_bagsize(t));

#endif // HAVE_GALA

}
