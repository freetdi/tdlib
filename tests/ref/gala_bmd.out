generated 129 edges, 32 vertices
symmetric 240 edges
0 -- 4
0 -- 7
0 -- 12
0 -- 21
0 -- 31
removed 0
treedec is valid!!
bagsize 15 status 0
boost_minDegree_ordering said 15
treedec is valid!!
